### ISSUE TITLE
Unregistered Handler, Ninject, .Net 4.8

### ISSUE THREAD

**Issue by WillMarcouiller** [jbogard/MediatR#664]
As per linked Ninject example:  [https://github.com/jbogard/MediatR/tree/master/samples/MediatR.Examples.Ninject](url), I guess I found an error in the [https://github.com/jbogard/MediatR/blob/master/samples/MediatR.Examples.Ninject/ContravariantBindingResolver.cs](url) class.

The line that I guess that is at fault is when getting the generic arguments for the second time.

    var genericType = service.GetGenericTypeDefinition();
    var genericArguments = genericType.GetGenericArguments();
    if (genericArguments.Count() == 1
       && genericArguments.Single().GenericParameterAttributes.HasFlag(GenericParameterAttributes.Contravariant))
    {
          var argument = service.GetGenericArguments().Single();
          var matches = bindings.Where(kvp => kvp.Key.IsGenericType
                                           && kvp.Key.GetGenericTypeDefinition().Equals(genericType)
                                           && kvp.Key.GetGenericArguments().Single() != argument
                                           && kvp.Key.GetGenericArguments().Single().IsAssignableFrom(argument))
                                .SelectMany(kvp => kvp.Value);
           return matches;
    }

Notice that `var genericArguments = genericType.GetGenericArguments()` uses the generic type definition returned by the `service.GetGenericTypeDefinition()` method call.  But because in the sample code provided the second call to `GetGenericArguments` is made from the `service` instance, the correct generic parameter doesn't seem to be returned adequately.  Hence I'd recommend the use of the already declared variable which contains the generic arguments from which the `argument` variable's value can be obtained.

To sum it up, for me what made the resolver actually resolve the correct handler every time is that I changed this line:

`var argument = service.GetGenericArguments().Single()`

to

`var argument = genericArguments.Single()`

given it is contravariant, so only available argument since the check is already made for the length of the returned arguments array.

Changing this for me made the difference between the unregistered handler exception to a working code able to resolve proper handlers.
